File: source/m-files/loadGATESPECTData.m
function [varargout] = loadGATESPECTData(varargin)
%LOADGATESPECTDATA Loads GATE SPECT projection images
%   Loads, up to 1000, GATE simulated SPECT projection images automatically.
%   Examples:
%       projData = loadGATESPECTData()
%       options = loadGATESPECTData(options)
%       [options, projData] = loadGATESPECTData(options)
%   Inputs:
%       options = The options-struct from a main-file (optional)
%   Outputs:
%       projData = The projection images as a matrix. Each slice (third
%       dimension) corresponds to a specific projection image. If multiple
%       heads are used, the data is loaded as GATE stores them, i.e. first
%       is head 1, then head 2, etc. The output is of type 16-bit unsigned
%       integer (uint16).
%       options = Optionally saves the size of the projection images, the
%       number of projections and the projection images in the options
%       struct.
%
% See also loadInterfile
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Copyright (C) 2022 Ville-Veikko Wettenhovi
%
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program. If not, see <https://www.gnu.org/licenses/>.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if isempty(varargin) || isempty(varargin{1}) || isempty(varargin{1}.fpath)
    [file, fpath] = uigetfile('*.sin','Select the first SPECT projection data file');
    if isequal(file, 0)
        error('No file was selected')
    end
else
    if exist(varargin{1}.fpath,'file') == 0
        warning('Specified file was not found! Please select first projection image')
        [file, fpath] = uigetfile('*.sin','Select the first SPECT projection data file');
        if isequal(file, 0)
            error('No file was selected')
        end
    else
        [fpath,file,~] = fileparts(varargin{1}.fpath);
    end
end
nimi = [fpath file];
nimi = [nimi(1:end-4), '.hdr'];
proj = cell(1000,1);
for kk = 1 : 1000
    try
        proj{kk} = loadInterfile(nimi);
    catch
        break
    end
    if kk < 10
        nimi = [nimi(1:end-5), num2str(kk + 1) '.hdr'];
    else
        nimi = [nimi(1:end-6), num2str(kk + 1) '.hdr'];
    end
end
loppu = kk - 1;
proj(cell2mat(cellfun(@isempty, proj, 'UniformOutput', false))) = [];
koko = cell2mat(cellfun(@size, proj, 'UniformOutput', false));
koko3 = max(koko(:,3));
A = zeros(koko(1,1), koko(1,2), koko3, class(proj{1}));
for kk = 1 : loppu
    cKoko = koko(kk,3);
    A(:,:,1:cKoko) = A(:,:,1:cKoko) + proj{kk};
end
if nargin > 0
    varargout{1} = varargin{1};
    varargout{1}.nRowsD = size(A,1);
    varargout{1}.nColsD = size(A,2);
%     varargout{1}.nProjections = size(A,3);
end
if nargout >= 2
    varargout{2} = A;
    varargout{1}.SinM = A;
elseif nargin == 0
    varargout{1} = A;
else
    varargout{1}.SinM = A;
end
end

File: source/Python/omegatomo/io/loadGATESPECTData.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar  7 13:34:05 2024

@author: Ville-Veikko Wettenhovi
"""
import numpy as np
import os
import tkinter as tk
from tkinter.filedialog import askopenfilename
from omegatomo.io.loadInterfile import loadInterfile

def loadGATESPECTData(options):
    if not os.path.isfile(options.fpath):
        print("Specified file was not found! Please select first projection image")
        root = tk.Tk()
        root.withdraw()
        options.fpath = askopenfilename(filetypes=[("SPECT projection data files", "*.sin")],
                                       title="Select the first SPECT projection data file")
        if not options.fpath:
            raise FileNotFoundError("No file was selected")
    fpath, file = os.path.split(options.fpath)
    fpath = os.path.abspath(fpath)

    nimi = os.path.join(fpath, file)
    nimi = nimi[:-4] + '.hdr'
    proj = [None] * 1000
    for kk in range(1000):
        try:
            proj[kk] = loadInterfile(nimi)
        except FileNotFoundError:
            break
        if kk < 10:
            nimi = nimi[:-5] + str(kk + 1) + '.hdr'
        else:
            nimi = nimi[:-6] + str(kk + 1) + '.hdr'
    loppu = kk

    proj = [p for p in proj if p is not None]
    koko = [p.shape for p in proj]
    koko3 = max([p[2] for p in koko])
    A = np.zeros((proj[0].shape[0], proj[0].shape[1], koko3), dtype=proj[0].dtype, order='F')
    for kk in range(loppu):
        cKoko = proj[kk].shape[2]
        A[:,:,:cKoko] += np.squeeze(proj[kk])
    options.nRowsD = A.shape[0]
    options.nColsD = A.shape[1]
    options.SinM = A
    return options